added optional configuration param 'translationsFile' that can be use…

treblereel opened the pull request:
…d to pass XTB translation file to closure compiler

niloc132 commented on build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:
```diff
@@ -141,6 +141,15 @@ public String getDependencyMode() {
     }
 
+    @Override
+    public Optional<String> getTranslationsFile() {
```
Please return a File instead of a String, and pass it through useFileConfig so that it is cached. Note that using a string in this way and not calling useStringConfig means that even if the string property value changed, the build would not re-run as it should, but in not using a file, if the file contents changed, the build still won't re-run.

niloc132 reviewed (commented):
Please add an integration test that confirms that this works, possibly building more than one locale if convenient.

treblereel commented:
hmm, looks like i have to revert my last commit (useFileConfig fix) coz it leads to:

```
[ERROR] Failed to execute goal com.vertispan.j2cl:j2cl-maven-plugin:0.20-SNAPSHOT:test (test-js) on project test: Failed to run tests: Bad artifact coordinates src/main/java/org/treblereel/test/MyTranslationBundle.xtb, expected format is <groupId>:<artifactId>[:<extension>[:<classifier>]]:<version> -> [Help 1]
```
 and it doesn't matter where i put MyTranslationBundle.xtb

niloc132 commented:
That error comes when a) the string is first used as a set of maven coords, and if that fails, b) the string is used as a path to pick out a local file. The idea is that you might want to point at an external artifact

```
        public File readFile() {
            String pathOrCoords = readString();
            if (pathOrCoords == null) {
                return null;
            }
            File f;
            // try to resolve as maven coords first
            try {
                f = getFileWithMavenCoords(pathOrCoords);
            } catch (ArtifactResolutionException e) {
                // handle as a file instead
                f = new File(pathOrCoords);
            }
            log.debug(getPath() + " => " + f.getAbsolutePath());
            return f;
        }
```

Probably you need to specify `${project.baseDir}/src/...` on your file, so that it is clear what the real path to the file is? I could see it being ambiguous, such as a generated xtb file from some other source, like merged properties files?

niloc132 commented on j2cl-maven-plugin/src/it/translation/src/main/java/org/treblereel/test/TranslationService.native.js:
```diff
@@ -0,0 +1,6 @@
+
+TranslationService.format = function () {
+    return MSG_SOMETHING_HAPPENED;
+}
```
semicolon, trailing newline

niloc132 commented on j2cl-maven-plugin/src/it/translation/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -0,0 +1,9 @@
+  version="2.5">
+  <!-- Default page to serve -->
+  <welcome-file-list>
+    <welcome-file>index.html</welcome-file>
```
if there is no actual index.html, this file is a little silly to have - take away the packaging=war, and the j2cl:build goal and just rely on the j2cl:test, then this file isnt needed?

treblereel replied:
done

niloc132 commented on j2cl-maven-plugin/src/it/translation/src/main/java/org/treblereel/test/holder.js:
```diff
@@ -0,0 +1,18 @@
+ */
+goog.provide('holder');
+/** @define {string} */
+holder.value = goog.define('holder.value', 'unknown');
```
trailing newline, copyright notice not required (following current repo conventions anyway - i might be wrong about it, but i'm not going to object if this file is copied without permission).

niloc132 commented on j2cl-maven-plugin/src/it/translation/src/main/java/org/treblereel/test/TranslationService.java:
```diff
@@ -0,0 +1,25 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
not required to put this in (and you seem inconsistent about where you add these?)

niloc132 commented on j2cl-maven-plugin/src/it/translation/pom.xml:
```diff
@@ -0,0 +1,16 @@
+    <module>pom_default.xml</module>
+    <module>pom_cz.xml</module>
+    <module>pom_fr.xml</module>
+  </modules>
```
consider one more project with three <execution> blocks, one for each test, each with its own <configuration> block to specify the expected file, path

niloc132 replied:
This is what I'm talking about, more than one execution in the same pom file, with different xtb file and local specified per execution - so that the user doesnt have to copy the whole pom file, just the execution.

niloc132 commented on build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:
```diff
@@ -141,6 +141,20 @@ public String getDependencyMode() {
+        }
+
+        try {
+            return Optional.of(new File(node.readString()).getCanonicalFile());
```
using readString means that only the string value of the path name is being tracked for changes - if the file itself is ever changed, only a clean rebuild will work (or deliberately change something unrelated)

treblereel commented:
@niloc132 thanks for youe reviews, now it works well with your https://github.com/Vertispan/j2clmavenplugin/pull/152 so i am going to rebase this branch with it once it ll be merged. 

* removed useless elemental2-dom
* packages renamed to example.test
* translation folder renamed to translationsFile
* fixed typos
* added more test cases

treblereel commented on build-caching/src/main/java/com/vertispan/j2cl/build/PropertyTrackingConfig.java:
```diff
@@ -141,6 +141,15 @@ public String getDependencyMode() {
+    @Override
+    public Optional<File> getTranslationsFile() {
+        ConfigValueProvider.ConfigNode node = config.findNode("translationsFile");
+        if (node == null) {
```
old school, forever old school :)

niloc132 commented on j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/JsZipBundleTask.java:
```diff
@@ -54,6 +54,7 @@ public Task resolve(Project project, Config config) {
                     Collections.emptyList(),
                     Collections.emptyMap(),
                     Collections.emptyList(),
+                    config.getTranslationsFile(),
```
This won't work I think, it should error out actually - did you confirm that BUNDLE_JAR behaves when there is a translations file?

niloc132 replied:
So two notes, first of all you can never make calls on `config` inside the lambda - it should error out (and i'm surprised it doesn't even with a non-empty translations file... worth looking into, that means nulls don't affect the cache), and second this is a BUNDLE so you should just give it Optional.empty()

niloc132 commented on j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/ClosureBundleTask.java:
```diff
@@ -96,6 +96,7 @@ public Task resolve(Project project, Config config) {
                     Collections.emptyList(),
                     Collections.emptyMap(),
                     Collections.emptyList(),//TODO actually pass these in when we can restrict and cache them sanely
+                    config.getTranslationsFile(),
```
as with JsZipBundleTask this shouldnt work, replace with Optional.empty()

niloc132 commented on j2cl-maven-plugin/src/main/java/com/vertispan/j2cl/mojo/TestMojo.java:
```diff
@@ -244,6 +244,9 @@
     @Parameter(defaultValue = "false")
     protected boolean enableSourcemaps;
 
+    @Parameter
```
docs? can follow up in a later release, but it should be the same as BuildMojo

niloc132 replied:
technically this property should be added to WatchMojo too, so that it is doc'd correctly, and can be set at the parent - see how languageOut, checkAssertions, rewritePolyfills are all declared but unused

treblereel replied:
are we going to run WatchMojo in advanced mode ? anyway, i ll add it

niloc132 replied:
Yes, that's a thing, I'm not going to tell people they _can't_ do it...

niloc132 replied:
Reading it now, it looks like you did miss WatchMojo?

niloc132 commented on j2cl-maven-plugin/src/main/java/com/vertispan/j2cl/mojo/BuildMojo.java:
```diff
@@ -134,6 +134,12 @@
     protected boolean rewritePolyfills;
 
+    /**
+     * Closure flag: "Source of translated messages. Currently only supports XTB."
```
also add a note that it will not work in BUNDLE or BUNDLE_JAR?

niloc132 commented on j2cl-tasks/src/main/java/com/vertispan/j2cl/tools/Closure.java:
```diff
@@ -112,6 +114,15 @@ public boolean compile(
         }
 
+        translationsFile.ifPresent(file -> {
+            if(compilationLevel.equals(CompilationLevel.ADVANCED_OPTIMIZATIONS)) {
```
stupid question perhaps, but are we sure this doesnt work with other modes?

...with that said, I dont think WHITESPACE_ONLY works at all right now, and I think SIMPLE is broken on windows

treblereel replied:
@niloc132 in this case we should create a check list of available modes and test PR againt them. I think, we can allow translationFiles to work with ADVANCED mode only.

niloc132 replied:
Sounds good, defer it to another issue?

treblereel replied:
yes, i think so

niloc132 commented:
I approved this, but please be sure to file that issue.
